### PR TITLE
Improve Web Driver Functional Test Coverage

### PR DESCRIPTION
**Pull request by xwsxethan**
I wanted to add more to these tests, but I realized this was already getting big, and I hit a stopping point with admin tests, so this seemed like a good place for a formal PR. The functional test additions and changes in this PR are as follows:
- New test suite for the search flows (from the landing and setup page) as well as the search page layout.
- Additional tests on the landing page for the disable/enable and create new invite code flows for a user.
- New test suite for the add admin flow. This will be expanded more in the future, but I realized the handler for removing an admin isn't wired up to any UI yet, so removing a test admin after adding is rather difficult.

I'm going to start another PR after this for wiring up a remove admin flow as well as unit tests for the admin/auth flow in general. That PR will then enable me to continue with the functional testing later.

Things still left to be functionally tested which are not covered here:
- Login/logout tests, verifying endpoints cannot be accessed while logged out.
- Settings tests, verifying that changing a setting works and propagates out to the chrome policy as necessary as well as verifying that each page redirects to the settings component.
- Repeated search tests, verifying that searching again after an initial search also works.
- Initial setup tests (need to figure out how to order these before other tests).
- Domain change tests (not sure if this will be possible since I don't think I'll be able to walk through the OAuth flow without having a Google user logged into the Chrome window).
- User add from domain tests, verifying that setting up a valid domain allows us to add users based on that domain, group, or user id.
- Error handler tests, verifying that the error handler catches various errors and displays them.
- Edit server tests, verifying that editing a server's properties actually gets respected.
- Tests for various services: key distributor actually dispenses keys, regexes actually stop invalid client input, user synchronizer actually gets updates from cron. Key distributor and user synchronizer may not be easily possible from an automated workflow.

---

This change is [Reviewable](https://reviewable.io/reviews/uproxy/ufo-management-server-flask/82)

### CODE REVIEW

File: tests/ui/admin_flow_test.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    login_url = self.args.server_url + flask.url_for('login')
+    self.assertEquals(login_url, self.driver.current_url)
+
+    # Login as an existing admin to get access.
+    LoginPage(self.driver).Login(self.args.server_url, self.args.username,
+                                 self.args.password)
+    self.driver.get(self.args.server_url + flask.url_for('landing'))
+
+    # Find the add admin form.
+    add_admin_form = self.getAdminFormHelper()
```
Comment (blueandgold) [inline tests/ui/admin_flow_test.py]: nit: might not have to use "helper" in the naming.  i.e. the way I read this line here, we are getting the form, not getting a helper
```python
add_admin_form = self.getAdminForm()
```

File: tests/ui/admin_flow_test.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    response_status = add_admin_form.find_element(
+        *UfOPageLayout.ADD_ADMIN_RESPONSE_STATUS)
+    self.assertIsNone(response_status)
+
+    # Add the new admin.
+    username_input = add_admin_form.find_element(
+        *UfOPageLayout.ADD_ADMIN_USERNAME)
+    username_real_input = username_input.find_element(By.ID, 'input')
```
Comment (blueandgold) [inline tests/ui/admin_flow_test.py]: nit: hmmm, I think the "username_real_input" here should actually be "username_input".  And above, maybe "username_input" should better be "username_row" (or something that describes what it is)????

File: tests/ui/admin_flow_test.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    username_real_input.send_keys(self.TEST_ADMIN_AS_DICT['username'])
+
+    password_input = add_admin_form.find_element(
+        *UfOPageLayout.ADD_ADMIN_PASSWORD)
+    password_real_input = password_input.find_element(By.ID, 'input')
```
Comment (blueandgold) [inline tests/ui/admin_flow_test.py]: ditto

File: tests/ui/landing_page_test.py [hunk 4]
```diff
@@ -42,75 +45,231 @@ def testLandingPageLayout(self):
...
-    """Manually add a test user using the landing page."""
-    # Navigate to add user and go to manual tab.
-    self.driver.get(self.args.server_url + flask.url_for('landing'))
-    landing_page = LandingPage(self.driver)
-    add_user_button = landing_page.GetElement(LandingPage.ADD_USER_BUTTON)
-    add_user_button.click()
-    add_manually_tab = WebDriverWait(self.driver, self.DEFAULT_TIMEOUT).until(
-        EC.visibility_of_element_located(((LandingPage.ADD_MANUALLY_TAB))))
-    add_manually_tab.click()
+  def enableDisableTestHelper(self, should_use_listbox, should_refresh_page):
```
Comment (blueandgold) [inline tests/ui/landing_page_test.py]: nit: Look at what this method is doing, maybe the naming here should be "disableAndThenEnableUser()".

**Comment by blueandgold** [comment]
:100:  This is awesome -- it's great to see our app being so well tested functionally!!

**Comment by xwsxethan** [comment]
Reviewed 14 of 16 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[tests/ui/admin_flow_test.py, line 52 \[r1\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/82#-KDiRsMfo2xmefW-9alB-r1-52:-KDihzOvxFAxRZSkILCO:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/6a73caca5d00ae2f50aa1fed88be853a15464de8/tests/ui/admin_flow_test.py#L52)):*
Done.

---

*[tests/ui/admin_flow_test.py, line 60 \[r1\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/82#-KDiRsMfo2xmefW-9alB-r1-60:-KDii-WnyHGz-zCxQrB-:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/6a73caca5d00ae2f50aa1fed88be853a15464de8/tests/ui/admin_flow_test.py#L60)):*
Done.

---

*[tests/ui/admin_flow_test.py, line 65 \[r1\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/82#-KDiRsMfo2xmefW-9alB-r1-65:-KDii0tfa0isU0NdMzO0:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/6a73caca5d00ae2f50aa1fed88be853a15464de8/tests/ui/admin_flow_test.py#L65)):*
Done.

---

*[tests/ui/landing_page_test.py, line 147 \[r1\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/82#-KDeiperOtVxOJyCe4Pu-r1-147:-KDii43h3UDzJVnnuyfc:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server-flask/blob/6a73caca5d00ae2f50aa1fed88be853a15464de8/tests/ui/landing_page_test.py#L147)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server-flask/82)*

**Pull request merged**